### PR TITLE
feat: adding compactor CLI command and crate

### PR DESCRIPTION
**Pull request by lukebond**
Skeleton of CLI command and crate for compactor. Mostly copied ingester and router2 here and there. Some TODO's left to indicate where to build out the bits to make it work.

Closes: #3777

### CODE REVIEW

File: compactor/src/handler.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+use async_trait::async_trait;
+use iox_catalog::interface::Catalog;
+use object_store::ObjectStore;
+use observability_deps::tracing::warn;
+use std::{fmt::Formatter, sync::Arc};
+use thiserror::Error;
+use tokio_util::sync::CancellationToken;
+
+#[derive(Debug, Error)]
+#[allow(missing_copy_implementations, missing_docs)]
```
Comment (mkmik) [inline compactor/src/handler.rs:12]: why `missing_copy_implementations`?

File: compactor/src/handler.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+pub enum Error {}
+
+/// The [`CompactorHandler`] does nothing at this point
+#[async_trait]
+pub trait CompactorHandler: Send + Sync {
+    /// Wait until the handler finished  to shutdown.
+    ///
+    /// Use [`shutdown`](Self::shutdown) to trigger a shutdown.
+    async fn join(&self);
```
Comment (mkmik) [inline compactor/src/handler.rs:21]: is it necessary that the CompactorHandler is Send+Sync?

what happens when multiple threads call `join` or shutdown?

Comment (crepererum) [reply compactor/src/handler.rs:21]: > what happens when multiple threads call join or shutdown?

The semantics are: `shutdown` is sync but doesn't block, it just flags the handler for shutdown. You may call this as often as you want. `join` blocks until the server finished shutdown (you can call this at any point and from as many threads as you wish).

**Review by crepererum — APPROVED** [review]

**Comment by lukebond** [comment]
thanks @crepererum for elucidating the stuff i copied over from ingester :laughing: :+1:

**Pull request merged**